### PR TITLE
[STAR] add hood open error

### PR DESCRIPTION
**Pull request by jkhales**
Debugged a confusing hood open error with @DonSqualo and wanted to make it easier for posterity.

### CODE REVIEW

File: pylabrobot/liquid_handling/backends/hamilton/STAR.py [hunk 1]
```diff
@@ -946,9 +956,7 @@ class STARFirmwareError(Exception):
   def __init__(self, errors: Dict[str, STARModuleError], raw_response: str):
     self.errors = errors
     self.raw_response = raw_response
-
-  def __repr__(self) -> str:
-    return f"{self.__class__.__name__}(errors={self.errors}, raw_response={self.raw_response})"
+    super().__init__(f"{errors}, {raw_response}")
```
Comment (jkhales) [inline pylabrobot/liquid_handling/backends/hamilton/STAR.py:959]: The error message doesn't display unless we do it in this more standard way

**Comment by rickwierenga** [comment]
Could you share a firmware log where this error occurred? It's interesting that the module is available but the `/` is missing in the error code. Or were you sending commands to I0 directly?

**Comment by jkhales** [comment]
> Could you share a firmware log where this error occurred? It's interesting that the module is available but the `/` is missing in the error code. Or were you sending commands to I0 directly?

I trigger this running `lh.setup()` with the hood open.

printing out right before this line: https://github.com/PyLabRobot/pylabrobot/blob/867b227d0b18b1701a235a84a542f211e8bb1250/pylabrobot/liquid_handling/backends/hamilton/STAR.py#L1184
```
errors_dict == {'I0': '36'} 
resp == I0XPid0006er36
```

full trace:

```
---------------------------------------------------------------------------
STARFirmwareError                         Traceback (most recent call last)
Input [In [2]](vscode-notebook-cell:?execution_count=2%3E), in <cell line: 5>()
      [2](vscode-notebook-cell:?execution_count=2&line=2) deck.assign_child_resource(resource=APER2506_CARRIER(Carrier.TALL.value), rails=19)
      [4](vscode-notebook-cell:?execution_count=2&line=4) lh = LiquidHandler(backend=STAR(), deck=deck)
----> [5](vscode-notebook-cell:?execution_count=2&line=5) await lh.setup()
      [7](vscode-notebook-cell:?execution_count=2&line=7) plate = Cos_96_EZWash("plate", with_lid=True)
      [8](vscode-notebook-cell:?execution_count=2&line=8) lh.deck.get_resource(Carrier.TALL.value).assign_resource_to_site(plate, 0)

File [c:\users\automation\pylabrobot\pylabrobot\liquid_handling\liquid_handler.py:117](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/liquid_handler.py:117), in LiquidHandler.setup(self)
    [114](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/liquid_handler.py:114) if self.setup_finished:
    [115](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/liquid_handler.py:115)   raise RuntimeError("The setup has already finished. See `LiquidHandler.stop`.")
--> [117](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/liquid_handler.py:117) await super().setup()
    [119](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/liquid_handler.py:119) self.head = {c: TipTracker(thing=f"Channel {c}") for c in range(self.backend.num_channels)}
    [120](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/liquid_handler.py:120) self.head96 = {c: TipTracker(thing=f"Channel {c}") for c in range(96)}

File [c:\users\automation\pylabrobot\pylabrobot\machine.py:63](file:///C:/users/automation/pylabrobot/pylabrobot/machine.py:63), in Machine.setup(self)
     [62](file:///C:/users/automation/pylabrobot/pylabrobot/machine.py:62) async def setup(self):
---> [63](file:///C:/users/automation/pylabrobot/pylabrobot/machine.py:63)   await self.backend.setup()
     [64](file:///C:/users/automation/pylabrobot/pylabrobot/machine.py:64)   self._setup_finished = True

File [c:\users\automation\pylabrobot\pylabrobot\liquid_handling\backends\hamilton\STAR.py:1245](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/STAR.py:1245), in STAR.setup(self)
   [1242](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/STAR.py:1242)   if not autoload_initialized:
   [1243](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/STAR.py:1243)     await self.initialize_autoload()
-> [1245](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/STAR.py:1245)   await self.park_autoload()
   [1247](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/STAR.py:1247) if self.iswap_installed:
   [1248](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/STAR.py:1248)   iswap_initialized = await self.request_iswap_initialization_status()

File [c:\users\automation\pylabrobot\pylabrobot\liquid_handling\backends\hamilton\STAR.py:5513](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/STAR.py:5513), in STAR.park_autoload(self)
   [5510](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/STAR.py:5510) max_x_pos = str(self.extended_conf["xt"]).zfill(2)
   [5512](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/STAR.py:5512) # Park autoload to max x position available
-> [5513](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/STAR.py:5513) return await self.send_command(
   [5514](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/STAR.py:5514)   module="I0",
   [5515](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/STAR.py:5515)   command="XP",
   [5516](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/STAR.py:5516)   xp=max_x_pos
   [5517](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/STAR.py:5517) )

File [c:\users\automation\pylabrobot\pylabrobot\liquid_handling\backends\hamilton\base.py:186](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/base.py:186), in HamiltonLiquidHandler.send_command(self, module, command, tip_pattern, write_timeout, read_timeout, wait, fmt, **kwargs)
    [167](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/base.py:167) """ Send a firmware command to the Hamilton machine.
    [168](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/base.py:168) 
    [169](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/base.py:169) Args:
   (...)
    [181](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/base.py:181)   A dictionary containing the parsed response, or None if no response was read within `timeout`.
    [182](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/base.py:182) """
    [184](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/base.py:184) cmd, id_ = self._assemble_command(module=module, command=command, tip_pattern=tip_pattern,
    [185](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/base.py:185)   **kwargs)
--> [186](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/base.py:186) resp = await self._write_and_read_command(id_=id_, cmd=cmd, write_timeout=write_timeout,
    [187](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/base.py:187)                 read_timeout=read_timeout, wait=wait)
    [188](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/base.py:188) if resp is not None and fmt is not None:
    [189](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/base.py:189)   return self._parse_response(resp, fmt)

File [c:\users\automation\pylabrobot\pylabrobot\liquid_handling\backends\hamilton\base.py:213](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/base.py:213), in HamiltonLiquidHandler._write_and_read_command(self, id_, cmd, write_timeout, read_timeout, wait)
    [211](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/base.py:211) fut = loop.create_future()
    [212](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/base.py:212) self._start_reading(id_, loop, fut, cmd, read_timeout)
--> [213](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/base.py:213) result = await fut
    [214](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/base.py:214) return cast(str, result)

File [c:\users\automation\pylabrobot\pylabrobot\liquid_handling\backends\hamilton\base.py:286](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/base.py:286), in HamiltonLiquidHandler._continuously_read(self)
    [284](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/base.py:284) if response_id == id_:
    [285](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/base.py:285)   try:
--> [286](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/base.py:286)     self.check_fw_string_error(resp)
    [287](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/base.py:287)   except Exception as e: # pylint: disable=broad-exception-caught
    [288](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/base.py:288)     loop.call_soon_threadsafe(fut.set_exception, e)

File [c:\users\automation\pylabrobot\pylabrobot\liquid_handling\backends\hamilton\STAR.py:1186](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/STAR.py:1186), in STAR.check_fw_string_error(self, resp)
   [1176](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/STAR.py:1176)   if error.message == "Unknown parameter":
   [1177](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/STAR.py:1177)     # temp. disabled until we figure out how to handle async in parse response (the
   [1178](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/STAR.py:1178)     # background thread does not have an event loop, and I'm not sure if it should.)
   (...)
   [1181](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/STAR.py:1181) 
   [1182](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/STAR.py:1182)     # pylint: disable=unnecessary-dict-index-lookup
   [1183](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/STAR.py:1183)     he.errors[module_name].message += \
   [1184](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/STAR.py:1184)       " (call lh.backend.request_name_of_last_faulty_parameter)"
-> [1186](file:///C:/users/automation/pylabrobot/pylabrobot/liquid_handling/backends/hamilton/STAR.py:1186) raise he

```

**Comment by rickwierenga** [comment]
The documentation says this is a "stop error" (for multiple modules defining error 36, but strangely not the C0 doc) so I renamed this error so it's re-usable for other modules in the future. If it is raised by I0 (autoload), then the trace information will specify the hood issue. `StopError(Hamilton will not run while the hood is open)`.

**Comment by rickwierenga** [comment]
Thank you for clarifying this error!

**Pull request merged**